torch/csrc/nnpack/NNPACK.cpp

```diff
@@ -0,0 +1,348 @@
+
+  // For now, we use the default algorithm
+  auto algorithm = nnp_convolution_algorithm_auto;
+
+  // Our input Tensor must be in the form N,C,H,W
+  if (input.ndimension() != 4) {
+    throw std::runtime_error("NNPack convolutionOutput expects 4D input Tensor N,C,H,W");
+  }
+  // Our weight Tensor must be in the form oC,iC,kH,kW
+  if (weight.ndimension() != 4) {
+    throw std::runtime_error("NNPack convolutionOutput expects 4D weight Tensor oC,iC,kH,kW");
+  }
+  // Our weight Tensor must be in the form N,oC,oH,oW
+  if (output.ndimension() != 4) {
+    throw std::runtime_error("NNPack convolutionOutput expects 4D output Tensor N,oC,oH,oW");
+  }
+
+  // All Tensors must be float Tensors
+  if (input.type().ID() != at::TypeID::CPUFloat ||
+      weight.type().ID() != at::TypeID::CPUFloat ||
+      output.type().ID() != at::TypeID::CPUFloat ||
+      (bias.defined() && bias.type().ID() != at::TypeID::CPUFloat)) {
+    throw std::runtime_error("Mismatched Tensor types in NNPack convolutionOutput");
+  }
+
+  const size_t batch_size = input.sizes()[0];
+  const size_t input_channels = input.sizes()[1];
+  const size_t output_channels = weight.sizes()[0];
+  const struct nnp_size input_size = {
+    .width = input.sizes()[3],
+    .height = input.sizes()[2]
+  };
+  const struct nnp_padding input_padding = {
+    .top = padH,
+    .right = padW,
+    .bottom = padH,
+    .left = padW
+  };
+  const struct nnp_size kernel_size = {
+    .width = kW,
+    .height = kH
+  };
+
+  // If we don't have a defined bias Tensor, we need to create one filled with zeroes
+  auto bias_ = bias.defined() ? bias : input.type().zeros({output_channels});
+
+  // Note: we assume that the output is shaped correctly, probably should add an assert
+
+  auto run = [&]() -> nnp_status {
+    return nnp_convolution_output(
+        algorithm,
+        batch_size,
+        input_channels,
+        output_channels,
+        input_size,
+        input_padding,
+        kernel_size,
+        (float*)input.data_ptr(),
+        (float*)weight.data_ptr(),
+        (float*)bias_.data_ptr(),
+        (float*)output.data_ptr(),
+        workspace, // workspace_buffer
+        &workspace_size, // workspace_size
+        nnp_activation_identity,
+        nullptr, // activation parameters
+        nnpack_threadpool(),
+        nullptr // profile
+    );
+  };
+
+  auto size_and_allocate_ws = [&]() {
+    // Run a single pass to get the size of memory workspace buffer
+    auto status = run();
+    if (status != nnp_status_success) {
+      throw std::runtime_error("NNPACK SpatialConvolution_updateOutput failed");
+    }
+    allocate_workspace();
+  };
+
+  // If no workspace created yet, allocate it
+  if (workspace == nullptr) {
+    size_and_allocate_ws();
+  }
+
+  // Try to run with the newly created, or existing workspace
+  auto status = run();
+
+  if (status == nnp_status_insufficient_buffer) {
+    // Need to reallocate the workspace
+    deallocate_workspace();
```

Comment:
This should be done automatically in `allocate_workspace()` right?

Reply:
Yes, but we have to deallocate the workspace prior to the call otherwise the kernel will simply return insufficient buffer space again instead of calculating the new size that we need.